### DISCUSSION TITLE
Loading data blob into buffer using C/C++

### DISCUSSION THREAD

**Discussion by riajain0412** [unicode-org/icu4x#3405] [Q&A]
I've tried loading the blob data into buffer using C as well as C++ . Below is the new C++ code which I've written:

int main(){
  UnicodeString str ("ພາສາລາວ");
  printUnicodeString(str);
  cout<<endl;

//C++ code
  std::ifstream fin("my_data_blob", std::ifstream::ate | std::ifstream::binary);
  fin.seekg(0, std::ios::end);
  std::streamsize blob_size = fin.tellg();
  cout<<"size buffer "<<blob_size<<endl;
  fin.seekg(0, std::ios::beg);
  char* buffer = new char[blob_size];
  fin.read(buffer, blob_size);
  cout<<buffer<<endl;
  std::vector<uint8_t> f_buffer(blob_size);
  auto sample = reinterpret_cast<uint8_t*>(buffer);
  auto sample2 = reinterpret_cast<uint8_t*>(buffer + blob_size);
  f_buffer.insert(f_buffer.end(), reinterpret_cast<const uint8_t*>(buffer), reinterpret_cast<const uint8_t*>(buffer+ blob_size));

  const diplomat::span<const uint8_t> blob(f_buffer.data(), f_buffer.size());
  const auto provider_result = ICU4XDataProvider::create_from_byte_slice(blob).ok().value();

  const auto segmenter_auto = ICU4XWordSegmenter::create_auto(provider_result).ok().value();
}
But this code is giving bad optional access error. Maybe there is some issue in loading blob into buffer. But this is the only way which I found to read blob's data into buffer. Are you aware of how to load your data blob into buffer?

**Comment by sffc** [comment]
An optional access error probably means your buffer is invalid. Your buffer reading code looks really complex so there's probably something wrong in there. However, that's a general C++ question, not an ICU4X question.

**Reply by riajain0412** [reply]
Yes, I know it's a general C++ question but I'm unable to figure out the way to read the data blob into buffer so I thought maybe you guys are aware of this. The code which I've written is the one and only way which I found on Google to read blob into buffer.

**Reply by riajain0412** [reply]
Do you have any idea about this?

**Reply by riajain0412** [reply]
Hi @sffc , Will you guys be able to help me out in this?

**Reply by sffc** [reply]
I can help answer questions involving the use of ICU4X. For questions outside that scope, I try to point you in the right direction, but this is not the right forum to seek out answers. Your question here is not about ICU4X but about how to read binary data into memory in C++.
